### ISSUE TITLE
API for adding custom modals or popups

### ISSUE THREAD

**Issue by sabieber** [PDFTron/webviewer-ui#34]
I really like the new API for adding custom panels to the left panel.
Could you add something similar for modals and popups so APIs like `closeElement` and `openElement` can be accessed there too?

**Comment by justinjung04** [comment]
Yup, we also think that similar API should exist for modals and popups. It's in our roadmap, but not the highest priority at the moment. Hopefully we will get to it soon and create better dev experience for everyone :)

**Comment by ascherer1993** [comment]
Something else I would like to add is that you might want to think about how you want to handle the onkey listeners for when a modal is open. In our product, we had a modal specifically linked to the selected annotation, and pressing delete would cause the annotation to be deleted and thus the modal to close. Because ondelete seemed to be handled in core code, I had to set up my own listener that would stop propagation if my modal was open. Perhaps you could turn off all listeners if a modal is open?

**Comment by justinjung04** [comment]
That's a great point. We'll consider it when we design the API.

**Comment by ghills** [comment]
This would be very useful for our use as well for page operations. For moving and adding pages we pop up separate modals right now but it is a bit clunky and doesn't work when using WebViewer in full screen mode.

**Comment by ZhijieZhang** [comment]
During this time we've added a few APIs that can be used to customize modals and popups:

- [instance.setCustomModal](https://www.pdftron.com/api/web/WebViewerInstance.html#setCustomModal) can be used to create a custom modal in the viewer. The modal can be toggled using `instance.openElements` and `instance.closeElements`.
- [Popup APIs](https://www.pdftron.com/api/web/WebViewerInstance.Popup.html) that can be used for `instance.annotationPopup`, `instance.contextMenuPopup` and `instance.textPopup`.
- The settings overlay can also be customized using `instance.settingsMenuOverlay`

Let us know if you have any other questions.

**Comment by sabieber** [comment]
@ZhijieZhang @justinjung04 any news on this?

**Comment by sabieber** [comment]
Is there any news on the Custom Popup API front?

**Comment by bollain** [comment]
Hi @sabieber,

We are currently working on a new and improved way to customize the UI without having to fork this repo, and can explore this requirement more in depth. 

If I understand correctly you want:
- An API to add a fully custom popup (not one of the default ones)
- A mechanism to open this based on an action (like a click on a certain annotation?)

Can you elaborate more on what your ideal use case is, and what an ideal API would look like for you?

**Comment by sabieber** [comment]
Hi @bollain,

thanks for the update and heads up! This sounds really promising.

Yes our use case is to create an annotation for the word the user clicked on and displaying a custom popup next to the created annotation.

Best would be an API where a popup with a fully custom content (like the dialog APIs) can be created. Also it would be helpful if there was an easy way to update the position of that popup to be next to a specified annotation.

**Comment by adam9000** [comment]
> Hi @sabieber,
> 
> We are currently working on a new and improved way to customize the UI without having to fork this repo, and can explore this requirement more in depth.
> 
> If I understand correctly you want:
> 
> * An API to add a fully custom popup (not one of the default ones)
> * A mechanism to open this based on an action (like a click on a certain annotation?)
> 
> Can you elaborate more on what your ideal use case is, and what an ideal API would look like for you?

Hi guys...

Can you give me, please, some hints on how this customization works?

Any help will be very much appreciated...It's a real struggle just to build it...

Even following those steps found in here ((https://docs.apryse.com/documentation/web/guides/advanced-customization/))

I only get to do this  after install Visual Studio 2019 with "Desktop development with C++" package...

Someone can help?

Tnx in advance!!!
